File: Cpp/1089_Duplicate_Zeros.cpp
class Solution {
public:
    void duplicateZeros(vector<int>& arr) {
        vector<int> duplicate(arr.size(), 0);
        
        for(int i=0, j=0; j<arr.size(); ++i, ++j){
            duplicate[j] = arr[i];
            if(!arr[i] && j+1<arr.size()){
                ++j;
                duplicate[j] = 0;
            }
        }
        
        arr = duplicate;
    }
};

File: Kotlin/1089_Duplicate_Zeros.kt
class Solution {
    fun duplicateZeros(arr: IntArray): Unit {
        var zeroCount = arr.count{it == 0}
        
        for(i in arr.size - 1 downTo 0){
            if(i + zeroCount <= arr.size - 1){
                arr[i + zeroCount] = arr[i]
                if(arr[i] == 0){
                    arr[i + zeroCount - 1] = 0
                    --zeroCount
                }
            } else {
                if(arr[i] == 0){
                    --zeroCount
                }
                if(i + zeroCount <= arr.size - 1){
                    arr[i + zeroCount] = arr[i]
                }
            }
        }
    }
}